### PR TITLE
Feat(connector): [WELLSFARGO] Implement Payment Flows

### PR DESCRIPTION
**Pull request by awasthi21**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description
Added Basic Payment flows for WellsFargo

### Additional Changes

- [x] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?
Added Cypress Test
**Some Tests are failing because of the 3ds request being routed to non_three_ds**
Screenshot 2024-08-05 at 5 32 51 PM

**For Ach Payments**
Request:
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key:' \
--data-raw '{
    "amount": 23243,
    "currency": "USD",
    "confirm": true,
    "capture_method": "automatic",
    "capture_on": "2022-09-10T10:11:12Z",
    "customer_id": "wellsfargocustomer",
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "phone_country_code": "+1",
    "return_url": "https://duck.com",
    "payment_method": "bank_debit",
    "payment_method_type": "ach",
    "payment_method_data": {
        "bank_debit": {
            "ach_bank_debit": {
                "name": "test",
                "bank_type":"checking",
                "account_number": "12345678",
                "routing_number": "112200439"
            }
        }
    },
    "billing": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "9123456789",
            "country_code": "+91"
        }
    },
    "shipping": {
        "address": {
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "city": "San Fransico",
            "state": "California",
            "zip": "94122",
            "country": "US",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "9123456789",
            "country_code": "+91"
        }
    }
}'
```
Response
```
{
    "payment_id": "pay_3jcprBJLvoggYKVGG1W3",
    "merchant_id": "postman_merchant_GHAction_65b8958b-ff29-4426-bc3c-c30f6fbd1946",
    "status": "succeeded",
    "amount": 23243,
    "net_amount": 23243,
    "amount_capturable": 0,
    "amount_received": 23243,
    "connector": "wellsfargo",
    "client_secret": "pay_3jcprBJLvoggYKVGG1W3_secret_rkW3Qfo9Xz58AVzW0ZnQ",
    "created": "2024-07-30T07:53:11.640Z",
    "currency": "USD",
    "customer_id": "cybersourcecustomer",
    "customer": {
        "id": "cybersourcecustomer",
        "name": "John Doe",
        "email": "guest@example.com",
        "phone": "999999999",
        "phone_country_code": "+1"
    },
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "bank_debit",
    "payment_method_data": {
        "bank_debit": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "9123456789",
            "country_code": "+91"
        },
        "email": null
    },
    "billing": {
        "address": {
            "city": "San Fransico",
            "country": "US",
            "line1": "1467",
            "line2": "Harrison Street",
            "line3": "Harrison Street",
            "zip": "94122",
            "state": "California",
            "first_name": "joseph",
            "last_name": "Doe"
        },
        "phone": {
            "number": "9123456789",
            "country_code": "+91"
        },
        "email": null
    },
    "order_details": null,
    "email": "guest@example.com",
    "name": "John Doe",
    "phone": "999999999",
    "return_url": "https://duck.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": null,
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "ach",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": {
        "customer_id": "wellsfargocustomer",
        "created_at": 1722325991,
        "expires": 1722329591,
        "secret": ""
    },
    "manual_retry_allowed": false,
    "connector_transaction_id": "7223259942386847504953",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": "pay_3jcprBJLvoggYKVGG1W3_1",
    "payment_link": null,
    "profile_id": "pro_2GK94kpxN56WZfJHV81h",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_vuiyOtLRLHEf1I97QC6p",
    "incremental_authorization_allowed": false,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-07-30T08:08:11.640Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-07-30T07:53:15.334Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null
}
```

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/wellsfargo.rs [hunk 1]
```diff
@@ -384,88 +1363,67 @@ impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::Payme
...
     }
 }
 
-impl ConnectorIntegration<api::Void, types::PaymentsCancelData, types::PaymentsResponseData>
-    for Wellsfargo
-{
-}
-
-impl ConnectorIntegration<api::Execute, types::RefundsData, types::RefundsResponseData>
+#[allow(dead_code)]
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wellsfargo.rs:1375]: Why is this added?

File: crates/router/src/connector/wellsfargo.rs [hunk 2]
```diff
@@ -308,19 +1248,49 @@ impl ConnectorIntegration<api::PSync, types::PaymentsSyncData, types::PaymentsRe
...
+        })
+    }
 }
 
-impl ConnectorIntegration<api::Capture, types::PaymentsCaptureData, types::PaymentsResponseData>
+impl api::Refund for Wellsfargo {}
+impl api::RefundExecute for Wellsfargo {}
+impl api::RefundSync for Wellsfargo {}
+
+#[allow(dead_code)]
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wellsfargo.rs:1287]: Why is this added?

File: crates/router/src/connector/wellsfargo.rs (hunk 1 above)
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wellsfargo.rs:1375]: Why is this added?

**Review by deepanshu-iiitu — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/wellsfargo.rs (hunk 2 above)
Comment (awasthi21) [reply crates/router/src/connector/wellsfargo.rs]: Not needed, Removed

File: crates/router/src/connector/wellsfargo.rs (hunk 1 above)
Comment (awasthi21) [reply crates/router/src/connector/wellsfargo.rs]: Not needed, Removed

**Review by deepanshu-iiitu — DISMISSED** [review]

File: crates/router/src/connector/wellsfargo/transformers.rs [hunk 3]
```diff
@@ -1,43 +1,1133 @@
...
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ApplePayPaymentInformation {
+    tokenized_card: TokenizedCard,
+}
+
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct MandatePaymentInformation {
+    payment_instrument: CybersoucrePaymentInstrument,
```
Comment (SamraatBansal) [inline crates/router/src/connector/wellsfargo/transformers.rs:374]: Please rename it to wellsFargo

File: crates/router/src/connector/wellsfargo/transformers.rs [hunk 4]
```diff
@@ -1,43 +1,1133 @@
...
+    Cards(Box<CardPaymentInformation>),
+    GooglePay(Box<GooglePayPaymentInformation>),
+    ApplePay(Box<ApplePayPaymentInformation>),
+    ApplePayToken(Box<ApplePayTokenPaymentInformation>),
+    MandatePayment(Box<MandatePaymentInformation>),
+    AchDebitPayment(Box<AchPaymentInformation>),
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CybersoucrePaymentInstrument {
```
Comment (SamraatBansal) [inline crates/router/src/connector/wellsfargo/transformers.rs:434]: Rename

File: cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js [hunk 5]
```diff
@@ -0,0 +1,492 @@
...
+                currency: "USD",
+                mandate_data: singleUseMandateData,
+            },
+            Response: {
+                status: 200,
+                body: {
+                    status: "succeeded",
+                },
+            },
+        },
```
Comment (pixincreate) [inline cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js:231]: Does `wellsFargo` support mandates? If not, you can remove.

File: cypress-tests/cypress/support/redirectionHandler.js [hunk 6]
```diff
@@ -295,7 +295,7 @@ function threeDsRedirection(redirection_url, expected_url, connectorId) {
         cy.get('input[type="password"]').type("password");
         cy.get("#buttonSubmit").click();
       });
-  } else if (connectorId === "bankofamerica" || connectorId === "cybersource") {
+  } else if (connectorId === "bankofamerica" || connectorId === "cybersource" || connectorId === "wellsfargo"){
```
Comment (pixincreate) [inline cypress-tests/cypress/support/redirectionHandler.js:298]: Please add a screenshot of tests running successfully.

File: cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js [hunk 7]
```diff
@@ -0,0 +1,492 @@
...
+            currency: "USD",
+        },
+    },
+};
+
+export const connectorDetails = {
+    card_pm: {
+        PaymentIntent: {
+            Request: {
+                payment_method_data: {
```
Comment (pixincreate) [inline cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js:55]: please add `payment_method: "card",` to all card exchanges (`Request`) as shown below. I've this [PR](https://github.com/juspay/hyperswitch/pull/5481) enforcing that change.

```suggestion
                payment_method: "card",
                payment_method_data: {
```

File: cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js (hunk 5 above)
Comment (awasthi21) [reply cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js:231]: Yes it does

File: cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js [hunk 8]
```diff
@@ -0,0 +1,492 @@
+const successfulNo3DSCardDetails = {
```
[the remark this answers is not in the document]
Comment (awasthi21) [reply cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js:1]: Attached

File: cypress-tests/cypress/support/redirectionHandler.js (hunk 6 above)
Comment (awasthi21) [reply cypress-tests/cypress/support/redirectionHandler.js:298]: Keeping this `{|| connectorId === "wellsfargo")}` for now, Three_DS flow is passing through non_three_ds flow

File: cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js (hunk 7 above)
Comment (awasthi21) [reply cypress-tests/cypress/e2e/PaymentUtils/WellsFargo.js:55]: Changed

File: crates/router/src/types/api.rs [hunk 9]
```diff
@@ -453,9 +453,9 @@ impl ConnectorData {
                 }
                 enums::Connector::Tsys => Ok(ConnectorEnum::Old(Box::new(&connector::Tsys))),
                 enums::Connector::Volt => Ok(ConnectorEnum::Old(Box::new(connector::Volt::new()))),
-                // enums::Connector::Wellsfargo => {
-                //     Ok(ConnectorEnum::Old(Box::new(connector::Wellsfargo::new())))
-                // }
+                enums::Connector::Wellsfargo => {
+                    Ok(ConnectorEnum::Old(Box::new(&connector::Wellsfargo)))
+                }
```
[the remark this answers is not in the document]
Comment (awasthi21) [reply crates/router/src/types/api.rs:458]: Taking this in next phase

**Review by SamraatBansal — APPROVED** [review]

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]
